# 판다스
## 판다스 시작- 파일을 DataFrame으로 로딩, 기본 API
## DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환
## DataFrame의 칼럼 데이터 세트 생성과 수정
## DataFrame 데이터 삭제
## Index 객체
## 데이터 셀렉션 및 필터링
## 정렬, Aggregation 함수, GroupBy 적용
## 결손 데이터 처리하기
## apply lamda 식으로 데이터 가공

## 1. 판다스 시작- 파일을 DataFrame으로 로딩, 기본 API
## 판다스의 핵심 객체: DataFrame

In [1]:
#pandas를 pd로 에일리어스(alias)해 임포트
import pandas as pd

#호출 시 파일명 인자로 들어온 파일을 로 딩해 DataFrame객체로 반환
titanic_df=pd.read_csv('titanic_train.csv')
titanic_df.head(3)

print('titanic 변수 type:', type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

In [2]:
#디폴트는 5, 맨 앞 3개의 로우만 표출
titanic_df.head(3)

#shape변수는 DataFrame의 행과 열을 튜플 형태로 반환
print('DataFrame 크기: ', titanic_df.shape)

#칼럼의 타입, NULL 데이터 개수, 데이터 분포 등 조회 가능
#info()메서드를 통해 총 데이터 건수, 데이터 타입, NULL건수를 알 수 있음
titanic_df.info()

#describe() 메서트를 통해 칼럼별 숫자형 데이터값의 n-percentile분포도, 평균값, 최댓값, 최솟값 나타냄. 오직 숫자형칼럼의 분포도만 조사
titanic_df.describe()

DataFrame 크기:  (891, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [3]:
#DataFrame의 []연산자 내부에 칼럼명을 입력하면 Series 형태로 반환
#value_counts 메서드: 해당 칼럼값의 유형과 건수 확인가능, 시리즈객체에서만 사용가능
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)



3    491
1    216
2    184
Name: Pclass, dtype: int64


Pclass 값 3이 491개, 1이 216개, 2가 184개 임을 의미. 많은 건수 순서로 정렬되어 값을 반환.
또한 이 결과 자체가 시리즈 객체. 앞의 3,1,2는 결과 시리즈의 index. 

## 2. DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환
## 넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기
### DataFrame은 2차원 데이터 이므로 2차원 이하의 데이터들만 DataFrame으로 변환될 수 있음!

In [4]:
import numpy as np

#먼저 1차원 데이터
col_name1=['col1']
list1=[1,2,3]
array1=np.array(list1)
print('array1 shape:', array1.shape)

#리스트를 이용해 DataFrame 생성
df_list1=pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame: \n', df_list1)

#넘파이 ndarray를 이용해 DataFrame 생성
df_array1=pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame: \n', df_array1)


array1 shape: (3,)
1차원 리스트로 만든 DataFrame: 
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame: 
    col1
0     1
1     2
2     3


In [5]:
#2차원 데이터
#3개의 칼럼명 필요함
col_name2=['col1','col2','col3']

#2행 3열의 리스트와 ndarray생성 후 이를 DataFrame으로 변환
list2=[[1,2,3],[11,12,13]]
array2=np.array(list2)
print('array2 shape:', array2.shape)
df_list2=pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame: \n', df_list2)
df_array2=pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame: \n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    12    13


In [6]:
#딕셔너리는 DataFrame으로 변환
#Key는 문자열 칼럼명으로 매핑, Value는 리스트형(또는 ndarray형) 칼럼 데이터로 매핑
dict={'col1':[1,11],'col2':[2,22],'col3':[3,33]}
df_dict=pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame: \n', df_dict)

딕셔너리로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    22    33


## DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기
### 많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용. 따라서 DataFrame을 다시 넘파이 ndarray로 변환하는 경우가 빈번하게 발생함.
### values를 이용한 ndarray로의 변환은 반드시 기억!!

In [7]:
#DataFrame을 ndarray로 변환
array3=df_dict.values
print('df_dict.values 타입:', type(array3),'df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [8]:
#DataFrame을 리스트로 변환
list3=df_dict.values.tolist()
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

#DataFrame을 딕셔너리로 변환
dict3=df_dict.to_dict('list')
print('\n df_dict.to dict() 타입:', type(dict3))
print(dict3)
#to_dict('list')인자로 list를 입력하면 딕셔너리의 값이 리스트형으로 반환

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


## 3. DataFrame의 칼럼 데이터 세트 생성과 수정

In [9]:
#타이타닉 데이터 프레임에 새로운 칼럼을 추가하고 일괄적으로 0 할당
titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [10]:
#기존 칼럼 Series의 데이터를 이용해 새로운 칼럼 Series 만들기
titanic_df['Age_by_10']=titanic_df['Age']*10
titanic_df['Family_No']=titanic_df['SibSp']+titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

In [11]:
#DataFrame 내의 기존 칼럼 값 업데이트
titanic_df['Age_by_10']=titanic_df['Age_by_10']+100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

## 4. DataFrame 데이터 삭제
## drop() 메서드원형: DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')
## labels, axis, inplace

In [12]:
#axis 0은 로우 방향 축, axis 1은 칼럼 방향 축
#drop() 메서드에 axis=1을 입력하면 칼럼 축 방향으로 드롭 수행
titanic_drop_df=titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

In [14]:
#inplace=True로 설정하면 원래 본인의 DataFrame의 데이터를 삭제
#여러 개의 컬럼을 삭제할 때 리스트 형태로 labels 파라미터 입력
drop_result=titanic_df.drop(['Age_0','Age_by_10','Family_No'],axis=1,inplace=True)
print('inplace=True 로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True 로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

반환된 값이 None. inplace를 트루로 설정한 채로 반환 값을 다시 자신의 데이터프레임 객체로 할당하지 말라는데 그럼 어떻게 ?? 모르겠음

In [15]:
#axis=0으로 설정해 index 로우 삭제
pd.set_option('display.width',100)
pd.set_option('display.max_colwidth',15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))
#index 0,1,2에 위치한 로우를 삭제
titanic_df.drop([0,1,2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket  \
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599   
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   

      Fare Cabin Embarked  
0   7.2500   NaN        S  
1  71.2833   C85        C  
2   7.9250   NaN        S  
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare  \
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000   
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   

  Cabin Embarked  
3  C123        S  
4   NaN        S  


인덱스 삭제 후 맨 앞 3개 데이터의 인덱스는 3,4,5이다. 
drop() 메서드에서 axis와 inplace인자를 적용해 dataframe을 변경하는 방식 정리
* axis: 로우를 삭제할 때는 axis=0, 칼럼을 삭제할 때는 axis=1로 설정
* 원본 DataFrame은 유지하고 드롭된 DataFrame을 새롭게 객체 변수로 받고 싶다면 inplace=False로 설정
* 원본 데이터 프레임에 드롭된 결과를 적용할 경우에는 inplace=True를 적용
* 원본 데이터프레임에서 드롭된 데이터프레임을 다시 원본 데이터프레임 객체 변수로 할당하면 원본 데이터프레임에서 드롭된 결과를 적용할 경우와 같음

## 5. Index 객체

In [16]:
#위에 drop하고 했으니 원본 파일 다시 로딩
titanic_df=pd.read_csv('titanic_train.csv')

#인덱스 객체 추출
indexes=titanic_df.index

#인덱스 객체를 실제 값 array로 변환, 인덱스 객체의 values 속성으로 ndarray값 알 수 있음
print('Index 객체 array값: \n', indexes.values)

Index 객체 array값: 
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241

In [17]:
#인덱스 객체는 식벽성 데이터를 1차원 array로 갖고 있음. 또한 ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


### 한번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없으며 오직 식별용으로만 사용되어 연산에서도 제외됨.

In [18]:
indexes[0]=5

TypeError: Index does not support mutable operations

In [19]:
series_fair=titanic_df['Fare']
print('Fair Sereis max 값:', series_fair.max())
print('Fair Sereis sum 값:', series_fair.sum())
print('sum() Fair Sereis:', sum(series_fair))
print('Fair Sereis +3:\n', (series_fair+3).head(3))

Fair Sereis max 값: 512.3292
Fair Sereis sum 값: 28693.9493
sum() Fair Sereis: 28693.949299999967
Fair Sereis +3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


In [20]:
#reset_index() 메서드를 수행하면 새롭게 인덱스를 할당하고 기존 인덱스는 데이터 프레임에 'index'라는 새로운 칼럼으로 추가됨
titanic_reset_df=titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  \
0      0            1         0       3  Braund, Mr....    male  22.0      1      0   
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0   
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0   

           Ticket     Fare Cabin Embarked  
0       A/5 21171   7.2500   NaN        S  
1        PC 17599  71.2833   C85        C  
2  STON/O2. 31...   7.9250   NaN        S

reset_index()는 인덱스가 연속된 int 숫자형이 아닐 경우에 주로 사용. 또한 시리즈 객체에 사용할 경우 데이터 프레임이 반환됨.

In [21]:
print('### before reset_index ###')
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))
new_value_counts=value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


## 데이터 셀렉션 및 필터링
## 정렬, Aggregation 함수, GroupBy 적용
## 결손 데이터 처리하기
## apply lamda 식으로 데이터 가공

과제-1
주제: 조원들을 더 알아보자...!!
1. 조원들의 이름, 학과, 다전공, 성별, 연도 등으로 데이터 프레임을 만들어 보기(3가지 방법 사용-ndarray, list, 딕셔너리를 데이터 프레임으로 변환하기)
2. 연도별로 오름차순 정렬하여 새로운 인덱스를 부여해 index 칼럼 만들고 나이순으로 번호메기기(hint:reset)
3. iloc[]연산자와 loc[] 연산자를 각각 활용해서 원하는 데이터 추출해보기(2개 이상)

과제-2
주제: 실습 고대로 해보기@
캐글에서 타이타닉 데이터 받아서 해보기